### DISCUSSION TITLE
Mutation of object type

### DISCUSSION THREAD

**Discussion by SteliosKornelakis** [pmndrs/valtio#525] [Q&A]
While it is clear from the documentation that we should not mess with the reference of the proxy parent itself 
-_If you reassign the proxy to an entirely new object, it will stop working._", it is not clear to me whether we should reassign any child of object type itself (by reference).

For example:
```
//My Store is:
const store = proxy(store:{id:1, data:{name:'Stelios'}})

//The parent component is something like this:
const DemoPage = () => {

    const store = GetStore();

    const click = () => {
        // store.data.name = 'Makis'; // This works
        store.data = { name: 'Makis' } // This does not work
    }

    return (
        <Container>
            <TestControl input={store} />
            <Button variant="outlined" onClick={click}>Outlined</Button>
        </Container>
    );
}

const TestControl = ({ store}: { store: ITestInput }) => {
    
    const unsubscribe = subscribe(store, () =>
        console.log('state has changed to', store) //This works
    )

    const { name } = useSnapshot(store.data) //This is not reactive !

    return (<>
        <TextField  value={name}  />
    </>)
}


```
 
So the question is when we have some big objects with many properties, is it required to change all properties down to the primitive types one by one instead of reassigning the reference of the object itself?

**Comment by dai-shi** [comment]
> when we have some big objects with many properties, is it required to change all properties down to the primitive types one by one instead of reassigning the reference of the object itself?

No, not always. It's only when you are subscribing to child objects (= "scoped subscriptions").
So, it depends on how you would use `useSnapshot`.

**Reply by SteliosKornelakis** [reply]
Well, I have some parent-child components which are using a derived-proxy of specific Objects (parts of the proxy Object tree).
These derived-proxy objects are being passed to components via props.

So there are cases where I subscribe to useSnapshot(parent.child1.child2).
While in the meantime I might need another component to parent.child1 = {}

I'm trying to understand if Valtios' mutable pattern maybe is not suitable for my case or I should just keep using it but mutating only the primitives of these objects.

**Reply by dai-shi** [reply]
> So there are cases where I subscribe to useSnapshot(parent.child1.child2).

Can you do this instead?
```js
useSnapshot(parent).child1.child2
```

**Reply by SteliosKornelakis** [reply]
I am doing something like this:
(My intention is to feed each component from the orchestrator/parent only with the required data)
```
const partOfStore = () => derive({
        store: (get) => {
            const a = {
                searchPlacementVm: get(parentStore.searchStore).placement,
                searchLocationVm: get(parentStore.searchStore).location,
            }
            return a;
        }
    })

//In the  respective components I 
const {..} = useSnapshot(placement) // in the placement component
const {..} = useSnapshot(location) // in the location component
```

However, searchStore could change have its reference changed during the lifecycle (from URL params for example).

I can change 
get(parentStore.searchStore).location to `get(parentStore).searchStore.location,`
however this will bring more rerenders due to unrelated props changes of the parent.

**Reply by dai-shi** [reply]
> however this will bring more rerenders due to unrelated props changes of the parent.

It will not, thanks to automatic render optimization.

**Reply by SteliosKornelakis** [reply]
I'll give it a try :)

**Reply by SteliosKornelakis** [reply]
Well, it works however, I cannot seem to avoid the unnecessary re-renders unless I just change the primitives down to their level.
When I pass to MasterComponent the derived state, on any sub-property it rerenders both controls.
```
const store = GetStore();

const derived: IStoreEntry<IHomeMenuSearchBarVm> = proxy({
    store: derive({
        searchPlacementVm: get => get(store).homeMenuSearchBarStore.store.placement,
        searchLocationVm: get => get(store).homeMenuSearchBarStore.store.location,
        searchFunc: get => (vm: IHomeMenuSearchBarVm) => { get(store).homeMenuSearchBarStore.actions.search.execute(vm) }
    }),
    actions: {}
});

//Dummy code
const MasterComponent(derived)=>{

return(<>
<UserControlA proxyState:{derived.store.placement} />
<UserControlA proxyState:{derived.store.location} />
</>)

}
```

**Reply by SteliosKornelakis** [reply]
I have managed to isolate the issue, but I think it's kind of by design.

MyStore:
```
const store = GetStore();

const derived: IStoreEntry<IHomeMenuSearchBarVm> = proxy({
    store: derive({
        searchPlacementVm: get => get(store).homeMenuSearchBarStore.store.placement,
        searchLocationVm: get => get(store).homeMenuSearchBarStore.store.location,
        searchFunc: get => (vm: IHomeMenuSearchBarVm) => { get(store).homeMenuSearchBarStore.actions.search.execute(vm) }
    }),
    actions: {}
});
```


The ParentContainer receives a derived object and only passes its two props (objects) to its children.

```
//Dummy code
const MasterComponent(derived)=>{

return(<>
<UserControlA proxyState:{derived.store.placement} />
<UserControlB proxyState:{derived.store.location} />
</>)
```


The thing is that since I do not use 'useSnaphot' on the parent when the derived.store object changes it does not trigger the update.
However, if I change only the primitives and since the references do not change, the useSnapshot on UserControlA and UserControlB are working as expected. 

I do not know if there is anything to be done besides adding 'useSnapshot' on the parent also. This would of course force both controls to rerender on each change of each

**Reply by dai-shi** [reply]
If you need help, I would like to help. But, I can't follow your code. The best is to create a _minimal_ reproduction with codesandbox to illustrate your issue.

**Reply by SteliosKornelakis** [reply]
Sure, I would appriciate your input on this. 
https://codesandbox.io/embed/quizzical-pike-vg7pef?fontsize=14&hidenavigation=1&theme=dark

Inspecting the console after you click the change non-primitive it will stop working.

**Reply by dai-shi** [reply]
It's still big. I think you can make it much smaller.
But, it looks like you are passing a part of `derived`, which causes the problem, right?
Why do you need `derived` in this use case? If you need it, you probably want to pass it all the way down.
Wish to have a smaller repro to make some modifications.

**Reply by SteliosKornelakis** [reply]
I'm afraid if I reduce it further we will not be able to reproduce.
I tried splitting it up a bit, though. (https://codesandbox.io/embed/quizzical-pike-vg7pef?fontsize=14&hidenavigation=1&theme=dark)

I am using the derived in order to pass only the required data to the components.
I am passing it all the way down.

The thing is that changing any non-primitive object of the store, for example:
```
store.userStore.obj1.name
derive({ ... get(store).userStore.obj1
```
Changing either userStore or obj1, the derived passed through props proxy is not updating.

**Reply by SteliosKornelakis** [reply]
I believe this would be easier to play :)
https://codesandbox.io/s/eager-cdn-odlsn3?file=/src/App.tsx

**Reply by dai-shi** [reply]
If you want to pass sub stores, `derive` isn't a good util for your case.
I would avoid using `derive` in this case, because it's confusing (like the behavior you observe.) 
Otherwise, the fix would be something like this: https://codesandbox.io/s/pedantic-tree-epprfu

**Reply by SteliosKornelakis** [reply]
Thank you for your time on this, I really appreciate it.
Nevertheless, I kindly find it a bit hackish for my taste :p.

I understand that this is the expected behavior as derive is not intended for such usage.
Well unless there is an alternative util like derive, I'm thinking of just proceeding with read-only non-primitives and just using a helper function for setting all properties of the objects - when required.
